cpp/velox/shuffle/VeloxShuffleWriter.cc

```diff
@@ -1134,24 +1152,29 @@ arrow::Status VeloxShuffleWriter::splitFixedWidthValueBuffer(const velox::RowVec
     return arrow::Status::OK();
   }
 
+  bool VeloxShuffleWriter::needRealloc(uint32_t partitionId, uint64_t newSize) {
+    auto avgBufferSize = partitionAvgBufferSize_[partitionId];
+    return avgBufferSize != 0 &&
+        ((newSize > (1 + options_.buffer_realloc_threshold) * avgBufferSize ||
+          newSize < (1 - options_.buffer_realloc_threshold) * avgBufferSize));
+  }
+
   uint32_t VeloxShuffleWriter::calculatePartitionBufferSize(const velox::RowVector& rv) {
     uint32_t sizePerRow = 0;
     auto numRows = rv.size();
     for (size_t i = fixedWidthColumnCount_; i < simpleColumnIndices_.size(); ++i) {
       auto index = i - fixedWidthColumnCount_;
-      if (binaryArrayEmpiricalSize_[index] == 0) {
-        auto column = rv.childAt(simpleColumnIndices_[i]);
-        auto stringViewColumn = column->asFlatVector<velox::StringView>();
-        assert(stringViewColumn);
-
-        // accumulate length
-        uint64_t length = stringViewColumn->values()->size();
-        for (auto& buffer : stringViewColumn->stringBuffers()) {
-          length += buffer->size();
-        }
-
-        binaryArrayEmpiricalSize_[index] = length % numRows == 0 ? length / numRows : length / numRows + 1;
+      auto column = rv.childAt(simpleColumnIndices_[i]);
+      auto stringViewColumn = column->asFlatVector<velox::StringView>();
+      assert(stringViewColumn);
+
+      // accumulate length
+      uint64_t length = stringViewColumn->values()->size();
+      for (auto& buffer : stringViewColumn->stringBuffers()) {
+        length += buffer->size();
       }
+
+      binaryArrayEmpiricalSize_[index] = (length + numRows - 1) / numRows;
```

Comment:
Why is not `binaryArrayEmpiricalSize_[index] = length % numRows == 0 ? length / numRows : length / numRows + 1;`

Reply:
it's just another implmentation

Reply:
we should count binary length using all history batches, instead of per rowvector. The value of each batch varies and still have the same issue